libtock/crt0.c

```diff
@@ -78,22 +78,25 @@ void _start(void* app_start __attribute__((unused)),
     "movs r5, #7\n"
     "bic  r4, r4, r5\n"         // r4 = (mem_start + myhdr->stack_size + 7) & ~0x7
     //
-    // Compute the heap size
+    // Compute the app data size and where initial app brk should go.
+    // This includes the GOT, data, and BSS sections. However, we can't be sure
+    // the linker puts them back-to-back, but we do assume they are in order:
+    // GOT then data then BSS (i.e. myhdr->got_start < myhdr->data_start <
+    // myhdr->bss_start). With all of that true, then the size is equivalent
```

Comment:
Tracking these bugs down is painful. Can we codify this assumption somewhere and make it part of the TBF standard? Perhaps enforced in the TBF parsing in the kernel or maybe tockloader (or both)

Reply:
This file would be the place to do it, or perhaps wherever we document that 10 word header. TBF explicitly does _not_ care about these details as apps are free to do whatever they want with their bits.

Reply:
Sigh.. I was afraid you were going to say that.

I think in this case we could probably get away with implementing this check in the `_c_start` function rather than in the assembly, that just needs any stack to be around to be safe, something a touch too small won't burn us there.

Reply:
Ok I added the check to the linker script, so if someone does try to move _got after _bss an error will show up.

Reply:
I agree that having it in the linker script is better. In general, I think our contract should be that you take the linker script and crt0 as a unit, and if you change either, you're responsible for both.

So specifically, the linker script in this repo, is a linker script that works with `libtock.a`'s `_start` function. You don't have to use that `_start` symbol if you choose to write your own linker script.